Support explicit defintion of resouce name for collection caching.

dubek opened the pull request:
If a template includes `# Template Collection: ...` anywhere in its
source, that name will be used as the cache name for the partial that is
rendered for the collection.

This allows users to enable collection caching even if the template
doesn't start with `<% cache ... do %>`.

Moreover, the `# Template Collection: ...` notation is recognized in all
template types (and template types other than ERB can define a
resource_cache_call_pattern method to allow the `cache ... do` pattern
to be recognized too).

This PR follows a discussion on #20417 .

@kaspth - appreciate your comments.

dubek committed:
Support explicit defintion of resouce name for collection caching.

If a template includes `# Template Collection: ...` anywhere in its
source, that name will be used as the cache name for the partial that is
rendered for the collection.

This allows users to enable collection caching even if the template
doesn't start with `<% cache ... do %>`.

Moreover, the `# Template Collection: ...` notation is recognized in all
template types (and template types other than ERB can define a
resource_cache_call_pattern method to allow the `cache ... do` pattern
to be recognized too).

kaspth commented on actionview/lib/action_view/template/handlers/erb.rb:
```diff
@@ -138,7 +138,14 @@ def call(template)
+            \s*                           # followed by optional spaces or newlines
+            (?<resource_name>\w+)         # capture the cache call argument as :resource_name
+            [\s\)]                        # followed by a space or close paren
+          /xm
```
Yes! I tried splitting the regex onto several lines originally, but couldn't get it to work. Thanks :heart:

dubek replied:
I ran into an issue with the `#` char - in normal regexp you don't need to escape it, but in multi-line (`/x`) regexp you need to escape it as `\#` , because Ruby comments are allowed *inside* these multi-line regexps. That was confusing.

kaspth replied:
Ha, I can imagine :smile:

kaspth commented:
Thanks for working on this! :heart:

We should update the documentation to say people can mark their templates. After this line might be a good place: https://github.com/dubek/rails/blob/collection-cache-explicit/actionview/lib/action_view/helpers/cache_helper.rb#L135.

We also need a changelog entry at the top of Action Views file :smile:

kaspth commented on actionview/lib/action_view/template.rb:
```diff
@@ -351,9 +351,21 @@ def instrument(action, &block)
+        if @handler.respond_to?(:resource_cache_call_pattern) &&
+           match = @source.match(@handler.resource_cache_call_pattern)
+          match[:resource_name]
+        end
```
How about this, :smile:

```ruby
def extract_resource_cache_name
  if match = @source.match(EXPLICIT_COLLECTION) || resource_cache_call_match
    match[:resource_name]
  end
end

def resource_cache_call_match
  if @handler.respond_to?(:resource_cache_call_pattern)
    @source.match(@handler.resource_cache_call_pattern)
  end
end
```

dubek commented:
Thanks for the comments. @kaspth - I accepted your small refactoring, and add CHANGELOG entry and documentation in cache_helper.rb. Squashed and rebased on master.

kaspth commented:
Looks great, @dubek. Thanks a lot for working on this :heart:

I'll just close and reopen for Travis to run again.
